qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,972 @@
+
+"""Class to store and manage the results of calibration experiments."""
+
+import os
+from collections import namedtuple, defaultdict
+from datetime import datetime
+from typing import Any, Dict, Set, Tuple, Union, List, Optional
+import csv
+import dataclasses
+import warnings
+import regex as re
+
+from qiskit.pulse import (
+    ScheduleBlock,
+    DriveChannel,
+    ControlChannel,
+    MeasureChannel,
+    Call,
+    Instruction,
+    AcquireChannel,
+    RegisterSlot,
+    MemorySlot,
+    Schedule,
+)
+from qiskit.pulse.channels import PulseChannel
+from qiskit.circuit import Parameter, ParameterExpression
+from qiskit_experiments.calibration.exceptions import CalibrationError
+from qiskit_experiments.calibration.parameter_value import ParameterValue
+
+ParameterKey = namedtuple("ParameterKey", ["parameter", "qubits", "schedule"])
+ScheduleKey = namedtuple("ScheduleKey", ["schedule", "qubits"])
+ParameterValueType = Union["ParameterExpression", float, int, complex]
+
+
+class Calibrations:
+    """
+    A class to manage schedules with calibrated parameter values. Schedules are
+    intended to be fully parameterized, including the index of the channels. See
+    the module-level documentation for extra details. Note that only instances of
+    ScheduleBlock are supported.
+    """
+
+    # The channel indices need to be parameterized following this regex.
+    __channel_pattern__ = r"^ch\d[\.\d]*\${0,1}[\d]*$"
+
+    def __init__(self, control_config: Dict[Tuple[int, ...], List[ControlChannel]] = None):
+        """Initialize the calibrations.
+
+        Args:
+            control_config: A configuration dictionary of any control channels. The
+                keys are tuples of qubits and the values are a list of ControlChannels
+                that correspond to the qubits in the keys.
+        """
+
+        # Mapping between qubits and their control channels.
+        self._controls_config = control_config if control_config else {}
+
+        # Store the reverse mapping between control channels and qubits for ease of look-up.
+        self._controls_config_r = {}
+        for qubits, channels in self._controls_config.items():
+            for channel in channels:
+                self._controls_config_r[channel] = qubits
+
+        # Dict of the form: (schedule.name, parameter.name, qubits): Parameter
+        self._parameter_map = {}
+
+        # Reverse mapping of _parameter_map
+        self._parameter_map_r = defaultdict(set)
+
+        # Default dict of the form: (schedule.name, parameter.name, qubits): [ParameterValue, ...]
+        self._params = defaultdict(list)
+
+        self._schedules = {}
+
+        # A variable to store all parameter hashes encountered and present them as ordered
+        # indices to the user.
+        self._hash_to_counter_map = {}
+        self._parameter_counter = 0
+
+    def add_schedule(self, schedule: ScheduleBlock, qubits: Union[int, Tuple[int, ...]] = None):
```

Comment:
Perhaps we can change this method name since this is no longer take `Schedule`. What about `add_calibration(self, calibration: ScheduleBlock, ...)`? Same for other method name `*_schedule`

Reply:
As discussed we shall leave as is.